Fill.py

```diff
@@ -129,21 +140,42 @@ def distribute_items_restrictive(world: MultiWorld, fill_locations=None):
         else:
             restitempool.append(item)
 
-    world.random.shuffle(fill_locations)
-    call_all(world, "fill_hook", progitempool, nonexcludeditempool, localrestitempool, nonlocalrestitempool, restitempool, fill_locations)
+    call_all(world, "fill_hook", progitempool, nonexcludeditempool,
+             localrestitempool, nonlocalrestitempool, restitempool, fill_locations)
+
+    locationDeficit = len(progitempool) - len(prioritylocations)
+    if locationDeficit > 0:
+        if locationDeficit > len(defaultlocations):
+            raise FillError(
+                f'Not enough locations for advancement items. There are {len(progitempool)} advancement items with {len(prioritylocations)} priority locations and {len(defaultlocations)} default locations')
+        prioritylocations += defaultlocations[:locationDeficit]
+        defaultlocations = defaultlocations[locationDeficit:]
+
+    fill_restrictive(world, world.state, prioritylocations, progitempool)
```

Comment:
If the locations in `prioritylocations` are not locked, then progression balancing will be able to move them earlier in the playthrough and they will no longer contain progression. Passing `lock=True` fixes this. (Credit to Alchav for noticing this.)

Reply:
Added a check to progression balancer that prevents items in PRIORITY locations from being rebalanced. There is definitely some conflict between the balancing  and priority. It seems like it would be beneficial to rework how the balancer works with this in mind at some point.